File: src/classes/mesh.cpp
#include "mesh.hpp"
#include <glm/glm.hpp>
#include <glm/gtc/type_ptr.hpp>
#include <glew.h>
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>
#include "camera.hpp"
#include "../config.hpp"

Mesh::Mesh (const std::string& filename, bool instanced, bool single, bool oriented) {
    loadFromFile(filename);
    setupMesh(instanced, single, oriented);
}

void Mesh::createSubVBO(GLuint &VBO, GLuint attributeIndex, GLsizei size, GLint numPerVertex, GLsizei stride, const void* pointer, const void* offset, GLuint divisor) {
    glGenBuffers(1, &VBO);
    glBindBuffer(GL_ARRAY_BUFFER, VBO);
    glBufferData(GL_ARRAY_BUFFER, size, pointer, GL_STREAM_DRAW);
    glEnableVertexAttribArray(attributeIndex);
    glVertexAttribPointer(attributeIndex, numPerVertex, GL_FLOAT, GL_FALSE, stride, offset);
    glVertexAttribDivisor(attributeIndex, divisor);
}

void Mesh::setupMesh(bool instanced, bool single, bool oriented) {
    glGenVertexArrays(1, &VAO);
    glGenBuffers(1, &VBO);
    glGenBuffers(1, &VBOPosition);

    glBindVertexArray(VAO);

    glBindBuffer(GL_ARRAY_BUFFER, VBO);
    glBufferData(GL_ARRAY_BUFFER, vertices.size() * sizeof(Vertex), &vertices[0], GL_STATIC_DRAW);

    glEnableVertexAttribArray(0);
    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex), (void*)0);

    glEnableVertexAttribArray(1);
    glVertexAttribPointer(1, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex), (void*)offsetof(Vertex, normal));

    glEnableVertexAttribArray(2);
    glVertexAttribPointer(2, 2, GL_FLOAT, GL_FALSE, sizeof(Vertex), (void*)offsetof(Vertex, texCoords));

    if (instanced) {
        if (!single) {
            createSubVBO(VBOPosition, 3, MAX_PARTICLES * sizeof(glm::vec3), 3, sizeof(glm::vec3), NULL, (void*)0, 1);

            createSubVBO(VBOscale, 4, MAX_PARTICLES * sizeof(glm::vec3), 3, sizeof(glm::vec3), NULL, (void*)0, 1);

            if (oriented) {
                createSubVBO(VBOrot, 5, MAX_PARTICLES * sizeof(glm::vec3), 3, sizeof(glm::vec3), NULL, (void*)0, 1);
            }

        } else { // case when we draw a single mesh

            createSubVBO(VBOPosition, 3, sizeof(glm::vec3), 3, sizeof(glm::vec3), NULL, (void*)0, 1);

            createSubVBO(VBOscale, 4, sizeof(glm::vec3), 3, sizeof(glm::vec3), NULL, (void*)0, 1);

            if (oriented) {
                createSubVBO(VBOrot, 5, sizeof(glm::vec3), 3, sizeof(glm::vec3), NULL, (void*)0, 1);
            }
        }
    }

    glBindBuffer(GL_ARRAY_BUFFER, 0);

    glBindVertexArray(0);
}

void Mesh::draw(GLuint& ShaderProgram, const Camera& camera, std::vector<glm::vec3>& positions, std::vector<glm::vec3>& scales) {
    glUseProgram(ShaderProgram);

    camera.loadMatricesIntoShader(ShaderProgram);

    glBindVertexArray(VAO);
    glBindBuffer(GL_ARRAY_BUFFER, VBOPosition);
    glBufferSubData(GL_ARRAY_BUFFER, 0, positions.size() * sizeof(glm::vec3), &positions[0]);

    glBindBuffer(GL_ARRAY_BUFFER, VBOscale);
    glBufferSubData(GL_ARRAY_BUFFER, 0, scales.size() * sizeof(glm::vec3), &scales[0]);

    glDrawArraysInstanced(GL_TRIANGLES, 0, vertices.size(), positions.size());
    glBindVertexArray(0);
}

void Mesh::drawOriented(GLuint& ShaderProgram, const Camera &camera, std::vector<glm::vec3>& positions, std::vector<glm::vec3>& scales, std::vector<glm::vec3>& rotations) {
    glUseProgram(ShaderProgram);

    camera.loadMatricesIntoShader(ShaderProgram);

    glBindVertexArray(VAO);
    glBindBuffer(GL_ARRAY_BUFFER, VBOPosition);
    glBufferSubData(GL_ARRAY_BUFFER, 0, positions.size() * sizeof(glm::vec3), &positions[0]);

    glBindBuffer(GL_ARRAY_BUFFER, VBOscale);
    glBufferSubData(GL_ARRAY_BUFFER, 0, scales.size() * sizeof(glm::vec3), &scales[0]);

    glBindBuffer(GL_ARRAY_BUFFER, VBOrot);
    glBufferSubData(GL_ARRAY_BUFFER, 0, rotations.size() * sizeof(glm::vec3), &rotations[0]);

    glDrawArraysInstanced(GL_TRIANGLES, 0, vertices.size(), positions.size());
    glBindVertexArray(0);

}

void Mesh::loadFromFile(const std::string& filename) {
    std::ifstream file(filename);
    std::string line;

    std::vector<glm::vec3> positions;
    std::vector<glm::vec2> texCoords;
    std::vector<glm::vec3> normals;

    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::string prefix;
        iss >> prefix;

        if (prefix == "v") {
            glm::vec3 position;
            iss >> position.x >> position.y >> position.z;
            positions.push_back(position);
        } else if (prefix == "vt") {
            glm::vec2 texCoord;
            iss >> texCoord.x >> texCoord.y;
            texCoords.push_back(texCoord);
        } else if (prefix == "vn") {
            glm::vec3 normal;
            iss >> normal.x >> normal.y >> normal.z;
            normals.push_back(normal);
        } else if (prefix == "f") {
            std::string vertexData[3];
            iss >> vertexData[0] >> vertexData[1] >> vertexData[2];

            for (const auto& data : vertexData) {
                std::istringstream viss(data);
                std::vector<int> indices(3);

                for (int& index : indices) {
                    viss >> index;
                    if (viss.peek() == '/') {
                        viss.ignore();
                    }
                }

                Vertex vertex;
                vertex.position = positions[indices[0] - 1];
                vertex.texCoords = texCoords[indices[1] - 1];
                vertex.normal = normals[indices[2] - 1];
                vertices.push_back(vertex);
            }
        }
    }
}

File: rust/src/classes/mesh.rs
extern crate glium;
extern crate nalgebra as na;

use glium::backend::Facade;
use glium::index::PrimitiveType;
use glium::vertex::VertexBuffer;
use glium::{implement_vertex, uniform, Program, Surface, Display, Frame};
use glium::draw_parameters::{DrawParameters};
use glutin::surface::{WindowSurface};
use na::{Vector2, Vector3};
use std::path::Path;
use std::ffi::CString;
use std::fs::File;
use std::io::{BufRead, BufReader};

use crate::classes::camera::{Camera};
use crate::config::{GLuint, MAX_PARTICLES, CAMERA_ASPECT_RATIO, CAMERA_FOV, CAMERA_NEAR, CAMERA_FAR};
use crate::classes::renderer::{InstancePosVertex, InstanceScaleVertex};

#[derive(Copy, Clone)]
struct Vertex {
    position: [f32; 3],
    normal: [f32; 3],
    tex_coords: [f32; 2],
}
implement_vertex!(Vertex, position, normal, tex_coords);

#[derive(Copy, Clone, Debug)]
pub struct SubVboVertexBuffer {
    pub data: [f32; 3],
}
implement_vertex!(SubVboVertexBuffer, data);

#[derive(Copy, Clone)]
pub struct Vec2Buffer {
    pub data: [f32; 2],
}
implement_vertex!(Vec2Buffer, data);

// ! important this has to be the exact names of the shader layouts
#[derive(Copy, Clone)]
pub struct ModelVertex {
    pub aPos: [f32; 3],
    pub aNormal: [f32; 3],
    pub aTexCoords: [f32; 2],
    pub instancePos: [f32; 3],
    pub instanceScale: [f32; 3],
}
implement_vertex!(ModelVertex, aPos, aNormal, aTexCoords, instancePos, instanceScale);

#[derive(Copy, Clone)]
pub struct APosVertex {
    pub aPos: [f32; 3],
}
implement_vertex!(APosVertex, aPos);

#[derive(Copy, Clone)]
pub struct ANormalVertex {
    pub aNormal: [f32; 3],
}
implement_vertex!(ANormalVertex, aNormal);

#[derive(Copy, Clone)]
pub struct ATexCoordsVertex {
    pub aTexCoords: [f32; 2],
}
implement_vertex!(ATexCoordsVertex, aTexCoords);

pub struct Mesh {
    // vao: glium::VertexArrayObject, // * automatically created by glium
    vbo: VertexBuffer<Vertex>,
    vbo_position: VertexBuffer<InstancePosVertex>,
    vbo_scale: VertexBuffer<InstanceScaleVertex>,
    vbo_rot: VertexBuffer<SubVboVertexBuffer>,
    vertices: Vec<Vertex>,
}

impl Mesh {
    pub fn new(display: &Display<WindowSurface>, filename: &str, instanced: bool, single: bool, oriented: bool) -> Self {
        let mut mesh = Mesh {
            vbo: VertexBuffer::empty_dynamic(display, 0).unwrap(),
            vbo_position: VertexBuffer::empty_dynamic(display, 0).unwrap(),
            vbo_scale: VertexBuffer::empty_dynamic(display, 0).unwrap(),
            vbo_rot: VertexBuffer::empty_dynamic(display, 0).unwrap(),
            vertices: Vec::new(),
        };
        mesh.load_from_file(filename);
        // mesh.setup_mesh(display, instanced, single, oriented);
        mesh
    }

    // pub fn create_sub_vbo<F: Facade>(vbo: &mut glium::VertexBuffer<Vector3<f32>>, attribute_index: u32, size: i32, num_per_vertex: i32, stride: i32, pointer: *const std::ffi::c_void, offset: *const std::ffi::c_void, divisor: u32, display: &F) {
    //     let mut data: Vec<SubVboVertexBuffer> = Vec::new();
    //     for i in 0..size {
    //         data.push(SubVboVertexBuffer { data: [0.0, 0.0, 0.0] });
    //     }
    //     glium::VertexBuffer::dynamic(display, &data).unwrap();
    // }

    fn extract_positions(&self) -> Vec<InstancePosVertex> {
        self.vertices.iter().map(|v| InstancePosVertex { instancePos: v.position }).collect()
    }

    fn extract_scales(&self) -> Vec<InstanceScaleVertex> {
        self.vertices.iter().map(|v| InstanceScaleVertex { instanceScale: v.normal }).collect()
    }

    // fn extract_rotations(&self) -> Vec<SubVboVertexBuffer> {
    //     self.vertices.iter().map(|v| SubVboVertexBuffer { data: v.tex_coords }).collect()
    // }

    pub fn setup_mesh(&mut self, display: &Display<WindowSurface>, instanced: bool, single: bool, oriented: bool) {
        if instanced {
            let max_particles = MAX_PARTICLES as usize; 

            if !single { // TODO check if usefull in the rust version
                self.vbo_position = VertexBuffer::new(display, &self.extract_positions()).unwrap();
                self.vbo_scale = VertexBuffer::new(display, &self.extract_scales()).unwrap();
                if oriented {
                    // self.vbo_rot = VertexBuffer::new(display, &self.extract_rotations()).unwrap();
                }
            } else {
                self.vbo_position = VertexBuffer::new(display, &self.extract_positions()).unwrap();
                self.vbo_scale = VertexBuffer::new(display, &self.extract_scales()).unwrap();


                if oriented {
                    // self.vbo_rot = VertexBuffer::new(display, &self.extract_rotations()).unwrap();
                }
            }
        }
    }

    pub fn draw(&mut self, shader_program: &Program, camera: &Camera, positions: &[InstancePosVertex], scales: &[InstanceScaleVertex], display: &Display<WindowSurface>, target: &mut Frame, params: &DrawParameters) {

        let uniforms = camera.get_uniforms(CAMERA_FOV, CAMERA_ASPECT_RATIO, CAMERA_NEAR, CAMERA_FAR);

        self.vbo_position = glium::VertexBuffer::new(display, positions).unwrap();
        self.vbo_scale = glium::VertexBuffer::new(display, scales).unwrap();

        // extract position, normal, and tex_coords from vertices
        let aPos = VertexBuffer::new(display, &self.vertices.iter().map(|v | APosVertex { aPos : v.position}).collect::<Vec<_>>()).unwrap();
        let aNormal = VertexBuffer::new(display, &self.vertices.iter().map(|v| ANormalVertex { aNormal : v.normal}).collect::<Vec<_>>()).unwrap();
        let aTexCoords = VertexBuffer::new(display, &self.vertices.iter().map(|v| ATexCoordsVertex { aTexCoords : v.tex_coords}).collect::<Vec<_>>()).unwrap();

        target.draw((&aPos, &aNormal, &aTexCoords, self.vbo_position.per_instance().unwrap(), self.vbo_scale.per_instance().unwrap()), glium::index::NoIndices(PrimitiveType::TrianglesList), shader_program, &uniforms, &params).unwrap();
        // target.draw((&aPos, &aNormal), glium::index::NoIndices(PrimitiveType::TrianglesList), shader_program, &uniforms, &Default::default()).unwrap();

    }

    pub fn draw_oriented<F: Facade>(&self, shader_program: &Program, camera: &Camera, positions: &[Vector3<f32>], scales: &[Vector3<f32>], rotations: &[Vector3<f32>], display: &F) {
        // Function body will be implemented later
        unimplemented!()
    }

    pub fn load_from_file(&mut self, filename: &str) {
        let file = match File::open(filename) {
            Ok(file) => file,
            Err(error) => panic!("Error opening file: {}", error),
        };
        let reader = BufReader::new(file);

        let mut positions = Vec::new();
        let mut tex_coords = Vec::new();
        let mut normals = Vec::new();

        for line in reader.lines() {
            let line = line.expect("Failed to read line");
            let mut parts = line.split_whitespace();
            let prefix = parts.next().unwrap_or("");

            match prefix {
                "v" => {
                    let x: f32 = parts.next().unwrap().parse().unwrap();
                    let y: f32 = parts.next().unwrap().parse().unwrap();
                    let z: f32 = parts.next().unwrap().parse().unwrap();
                    positions.push(Vector3::new(x, y, z));
                }
                "vt" => {
                    let x: f32 = parts.next().unwrap().parse().unwrap();
                    let y: f32 = parts.next().unwrap().parse().unwrap();
                    tex_coords.push(Vector2::new(x, y));
                }
                "vn" => {
                    let x: f32 = parts.next().unwrap().parse().unwrap();
                    let y: f32 = parts.next().unwrap().parse().unwrap();
                    let z: f32 = parts.next().unwrap().parse().unwrap();
                    normals.push(Vector3::new(x, y, z));
                }
                "f" => {
                    let vertex_data: Vec<&str> = parts.collect();
                    for data in vertex_data {
                        let indices: Vec<&str> = data.split('/').collect();
                        let position_index: usize = indices[0].parse().unwrap();
                        let tex_coord_index: usize = indices[1].parse().unwrap();
                        let normal_index: usize = indices[2].parse().unwrap();

                        let vertex = Vertex {
                            position: positions[position_index - 1].into(),
                            tex_coords: tex_coords[tex_coord_index - 1].into(),
                            normal: normals[normal_index - 1].into(),
                        };
                        self.vertices.push(vertex);
                    }
                }
                _ => {}
            }
        }
    }

}